File: src/rng.js
// =============================================================================
// Arcade — seedable deterministic RNG
// =============================================================================
// The keystone of no-cloud multiplayer: a PRNG seeded from a shared value (a
// room code / pairing session) produces the SAME sequence on every device, so
// two phones shuffle the same deck without a server refereeing. Replacing each
// game's ad-hoc Math.random() with one seedable source also makes games
// reproducible — record the seed + actions and you can replay a match exactly.
//
// mulberry32: tiny, fast, good distribution for game use (not cryptographic).
// =============================================================================

/** FNV-1a hash of a string → uint32 seed. */
export function hashSeed(value) {
  let h = 2166136261 >>> 0;
  const s = String(value);
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  return h >>> 0;
}

/**
 * Create a deterministic RNG from a seed (string or uint32).
 * @param {string|number} seed
 */
export function makeRng(seed) {
  let a = typeof seed === 'number' ? seed >>> 0 : hashSeed(seed);

  const next = () => {
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };

  const rng = {
    /** float in [0, 1). */
    next,
    /** integer in [0, n). */
    int: (n) => Math.floor(next() * n),
    /** integer in [min, max] inclusive. */
    range: (min, max) => min + Math.floor(next() * (max - min + 1)),
    /** a random element (or null). */
    pick: (arr) => (arr.length ? arr[Math.floor(next() * arr.length)] : null),
    /** a new array, Fisher–Yates shuffled. Does not mutate the input. */
    shuffle: (arr) => {
      const out = [...arr];
      for (let i = out.length - 1; i > 0; i--) {
        const j = Math.floor(next() * (i + 1));
        [out[i], out[j]] = [out[j], out[i]];
      }
      return out;
    },
    /** fork a child RNG (independent stream) for sub-systems. */
    fork: (label) => makeRng((a ^ hashSeed(String(label))) >>> 0),
  };
  return rng;
}

File: go/rng/rng.go
// Package rng is Arcade's seedable deterministic PRNG (mulberry32).
//
// It is byte-identical to the JavaScript @ric/arcade rng: same FNV-1a seed
// hash, same mulberry32 step. So a Go client and a JS client that seed from the
// same value (e.g. a waves_worx pairing session) produce the same shuffle and
// can play lockstep multiplayer ACROSS languages. Not cryptographic.
package rng

// HashSeed is FNV-1a over a string → uint32 (matches JS hashSeed).
func HashSeed(s string) uint32 {
	h := uint32(2166136261)
	for i := 0; i < len(s); i++ {
		h ^= uint32(s[i])
		h *= 16777619
	}
	return h
}

// RNG is a mulberry32 generator. Use New / NewUint; not safe for concurrent use.
type RNG struct{ a uint32 }

// New seeds from a string.
func New(seed string) *RNG { return &RNG{a: HashSeed(seed)} }

// NewUint seeds from a uint32 directly.
func NewUint(seed uint32) *RNG { return &RNG{a: seed} }

// Next returns a float64 in [0, 1). uint32 arithmetic wraps mod 2^32, matching
// JS Math.imul + >>>0 exactly.
func (r *RNG) Next() float64 {
	r.a += 0x6d2b79f5
	t := (r.a ^ (r.a >> 15)) * (1 | r.a)
	t = ((t + ((t ^ (t >> 7)) * (61 | t))) ^ t)
	return float64(t^(t>>14)) / 4294967296.0
}

// Int returns an int in [0, n).
func (r *RNG) Int(n int) int { return int(r.Next() * float64(n)) }

// Range returns an int in [min, max] inclusive.
func (r *RNG) Range(min, max int) int { return min + int(r.Next()*float64(max-min+1)) }

// Shuffle returns a new Fisher–Yates-shuffled copy (input untouched), matching
// the JS rng.shuffle order for the same seed.
func Shuffle[T any](items []T, r *RNG) []T {
	out := make([]T, len(items))
	copy(out, items)
	for i := len(out) - 1; i > 0; i-- {
		j := int(r.Next() * float64(i+1))
		out[i], out[j] = out[j], out[i]
	}
	return out
}
